File: projects/scala.emac.sh/src/example/99BottlesOfBear.scala
package example

import scala.Range

object Bear {
  def main(args: Array[String]) {
    var i = 0
    for (i <- Range(99, 2, -1)) {
      println(s"$i bottles of beer on the wall, $i bottles of beer.\nTake one down and pass it around, ${i - 1} bottles of beer on the wall.\n")
    }
    i = 2
    println(s"$i bottles of beer on the wall, $i bottles of beer.\nTake one down and pass it around, 1 bottle of beer on the wall.\n")
    i -= 1
    println(s"$i bottle of beer on the wall, $i bottle of beer.\nTake one down and pass it around, no more bottles of beer on the wall.\n")
    println("No more bottles of beer on the wall, no more bottles of beer.\nGo to the store and buy some more, 99 bottles of beer on the wall.")
  }
}

File: playground/python/99bottlesofbear.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-


# Non-OOP
def get_words(idx):
    if idx > 1:
        if idx > 2:
            tail = '%s bottles' % idx
        else:
            tail = '1 bottle'
        return ["%s bottles" % idx, "%s bottles" % idx, 'Take one down and pass it around', tail]
    elif idx == 1:
        return ['1 bottle', '1 bottle', 'Take one down and pass it around', "no more bottles"]
    else:
        return ["No more bottles", 'no more bottles', 'Go to the store and buy some more', "99 bottles"]


for i in range(99, -1, -1):
    words = get_words(i)
    print "%s of beer on the wall, %s of beer.\n%s, %s of beer on the wall.\n" % (
        words[0], words[1], words[2], words[3])


# OOP
class Bottle:
    @staticmethod
    def get_instance(idx):
        if idx > 1:
            return NBottle(idx)
        elif idx == 1:
            return OneBottle(idx)
        else:
            return ZeroBottle(idx)

    def __init__(self, idx):
        self.idx = idx

    def sing(self):
        words = self.get_words()
        print "%s of beer on the wall, %s of beer.\n%s, %s of beer on the wall.\n" % (
            words[0], words[1], words[2], words[3])


class NBottle(Bottle):
    def get_words(self):
        idx = self.idx
        if idx > 2:
            tail = '%s bottles' % idx
        else:
            tail = '1 bottle'
        return ["%s bottles" % idx, "%s bottles" % idx, 'Take one down and pass it around', tail]


class OneBottle(Bottle):
    def get_words(self):
        return ['1 bottle', '1 bottle', 'Take one down and pass it around', "no more bottles"]


class ZeroBottle(Bottle):
    def get_words(self):
        return ["No more bottles", 'no more bottles', 'Go to the store and buy some more', "99 bottles"]


for i in range(99, -1, -1):
    Bottle.get_instance(i).sing()
